lib/sse_test_server/request_handler.ex

```diff
@@ -0,0 +1,140 @@
+defmodule SSETestServer.RequestHandler do
+  @moduledoc """
+  HTTP request handler for everything.
+
+  TODO: Find a way to get rid of the chunk wrappers around all this stuff to
+  allow testing clients that don't handle transport-encodings transparently.
+  """
+
+  @behaviour :cowboy_loop
+
+  alias SSETestServer.SSEServer
+
+  defmodule State do
+    @enforce_keys [:sse_server]
+    defstruct sse_server: nil, path: nil
+  end
+
+  defmodule StreamState do
+    @enforce_keys [:path, :sse_server]
+    defstruct path: nil, sse_server: nil, response_delay: nil
+  end
+
+  def init(req = %{method: "PUT"}, state) do
+    {:ok, field_list, req_read} = :cowboy_req.read_urlencoded_body(req)
+    handle_put_endpoint(Map.new(field_list), req_read, state)
+  end
+
+  def init(req = %{method: "GET"}, state) do
+    when_exists(req, state, fn endpoint ->
+      case :cowboy_req.parse_header("accept", req) do
+        [{{"text", "event-stream", _}, _, _}] ->
+          handle_sse_stream(req, state, endpoint.handler_state)
+        _ -> {:ok, :cowboy_req.reply(406, req), state}
+      end
+    end)
+  end
+
+  def init(req = %{method: "POST"}, state) do
+    when_exists(req, state, fn _ ->
+      {:ok, field_list, req_read} = :cowboy_req.read_urlencoded_body(req)
+      handle_action(Map.new(field_list), req_read, state)
+    end)
+  end
+
+  def handle_sse_stream(req, state, sse_state) do
+    :ok = SSEServer.sse_stream(state.sse_server, req.path, self())
+    # sse_state.response_delay is nil (falsey) or an integer (truthy).
+    if sse_state.response_delay, do: Process.sleep(sse_state.response_delay)
+    req_resp = :cowboy_req.stream_reply(
+      200, %{"content-type" => "text/event-stream"}, req)
+    {:cowboy_loop, req_resp, sse_state}
+  end
+
+  defp when_exists(req, state, fun) do
+    case SSEServer.get_endpoint(state.sse_server, req.path) do
+      {:ok, endpoint} -> fun.(endpoint)
+      :error -> {:ok, :cowboy_req.reply(404, req), state}
+    end
+  end
+
+  defp handle_put_endpoint(fields, req, state) do
+    handler_opts =
+      [fn -> get_handler_opt(fields, :response_delay, &String.to_integer/1) end]
+      |> Enum.flat_map(&apply(&1, []))
+    SSEServer.add_endpoint(state.sse_server, req.path, handler_opts)
+    {:ok, :cowboy_req.reply(201, req), state}
```

Comment:
It will replace the existing one. We should probably handle that as a reconfiguration (keeping the existing streams but replacing the handler state) and add a deletion API. That's probably a thing for a new branch, though.

Reply:
What happens to clients connected to the old one?

Reply:
I believe the current behaviour (which I haven't tested) is that existing requests remain connected but don't receive any further messages from `SSEServer`. Any new requests (even those reusing existing TCP connections) will go to the new handler.

Reply:
I decided to fix the reconfiguration in this PR (because that's arguably a bug) but leave the deletion for later.